### PR TITLE
Signal live streaming with 'live' event

### PR DESCRIPTION
**Pull request by robin-da**
This adds a new 'live' event to our streams that signals that the WebSocket
streams are now live updating. It is emitted as soon as a `{events: [], offset:
n}` message is received from the WebSocket.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: language-support/ts/daml-react/index.test.ts [hunk 1]
```diff
@@ -351,6 +387,9 @@ describe('useStreamQuery', () => {
     mockStreamQuery.mockClear();
     mockStreamQuery.mockReturnValueOnce(stream);
     act(() => result.current.setQuery(query2));
+    // live event
+    act(() => void emitter.emit('live', []));
```
Comment (hurryabit) [inline language-support/ts/daml-react/index.test.ts:391]: `[]` is not a valid state for a streaming fetch-by-key.
```suggestion
    act(() => void emitter.emit('live', null));
```

File: language-support/ts/daml-react/hooks.ts [hunk 2]
```diff
@@ -154,12 +154,12 @@ export function useStreamQuery<T extends object, K, I extends string>(template:
     const query = queryFactory ? queryFactory() : undefined;
     console.debug(`mount useStreamQuery(${template.templateId}, ...)`, query);
     const stream = state.ledger.streamQuery(template, query);
+    stream.on('live', contracts => setResult(result => ({...result, contracts, loading: false})));
```
[the remark this answers is not in the document]
Comment (robin-da) [reply language-support/ts/daml-react/hooks.ts]: so should the type of the 'live' handler not get the state as input?

File: language-support/ts/daml-ledger/index.ts [hunk 3]
```diff
@@ -192,6 +192,7 @@ export type StreamCloseEvent = {
  * @typeparam State The accumulated state.
  */
 export interface Stream<T extends object, K, I extends string, State> {
+  on(type: 'live', listener: (state: State) => void): void;
```
[the remark this answers is not in the document]
Comment (robin-da) [reply language-support/ts/daml-ledger/index.ts:195]: added.

File: language-support/ts/daml-react/hooks.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (robin-da) [reply language-support/ts/daml-react/hooks.ts]: ah, you answered my question above.

File: language-support/ts/daml-ledger/index.ts [hunk 4]
```diff
@@ -502,12 +503,8 @@ class Ledger {
         const events = jtv.Result.withException(jtv.array(decodeEvent(template)).run(json.events));
         if (events.length == 0 && isRecordWith('offset', json)) {
           if (!haveSeenEvents) {
-            // NOTE(MH): If we receive the marker indicating that we are switching
-            // from the ACS to the live stream and we haven't received any events
-            // yet, we signal this by pretending we received an empty list of
-            // events. This never does any harm.
             haveSeenEvents = true;
```
[the remark this answers is not in the document]
Comment (robin-da) [reply language-support/ts/daml-ledger/index.ts]: I changed the `streamSubmit` method accordingly.

File: language-support/ts/daml-ledger/index.test.ts [hunk 5]
```diff
@@ -5,19 +5,20 @@ import { Template, Choice, ContractId } from "@daml/types";
 import Ledger, {CreateEvent} from "./index";
 import { Event } from "./index";
 import * as jtv from "@mojotech/json-type-validation";
-import type { EventEmitter } from 'events';
+import { EventEmitter } from 'events';
```
Comment (hurryabit) [inline language-support/ts/daml-ledger/index.test.ts:8]: Why? The changes below don't need this.

File: language-support/ts/daml-ledger/index.ts [hunk 6]
```diff
@@ -500,21 +502,11 @@ class Ledger {
...
-          state = change(state, events);
-          if (isRecordWith('offset', json)) {
-            haveSeenEvents = true;
-          }
-          emitter.emit('change', state, events);
+        state = change(state, events);
+        emitter.emit('change', state, events);
+        if (isRecordWith('offset', json) && !isLive) {
+          isLive = true;
+          emitter.emit('live', state);
```
Comment (hurryabit) [inline language-support/ts/daml-ledger/index.ts:509]: That is exactly what I had in mind. Awesome!

File: language-support/ts/daml-ledger/index.ts [hunk 7]
```diff
@@ -500,21 +502,11 @@ class Ledger {
...
-            emitter.emit('change', state, []);
-          }
-        } else {
-          state = change(state, events);
-          if (isRecordWith('offset', json)) {
-            haveSeenEvents = true;
-          }
-          emitter.emit('change', state, events);
+        state = change(state, events);
+        emitter.emit('change', state, events);
```
Comment (hurryabit) [inline language-support/ts/daml-ledger/index.ts:506]: We don't want to trigger any changes if there were no events. This happens as part of heart beating and informing about new ledger offsets.
```suggestion
        if (events.length > 0) {
          state = change(state, events);
          emitter.emit('change', state, events);
        }
```

File: language-support/ts/daml-ledger/index.test.ts (hunk 5 above)
Comment (robin-da) [reply language-support/ts/daml-ledger/index.test.ts:8]: that was just for me when debugging. somehow my IDE doesn't like this. I'll change it back when ready.

**Pull request merged**